repository/client.go

```diff
@@ -281,92 +307,131 @@ func (sc secretClient) UpdateClientData(ctx context.Context, update clientData)
-
 	go func() {
-		defer close(resultChan)
-		for {
-			select {
-			case <-ctx.Done():
-				watcher.Stop()
-				return
-			case result, open := <-watcher.ResultChan():
-				if !open {
-					return
-				}
-
-				resultChan <- handleWatchEvent(name, result)
-			}
-		}
+		informer.Run(ctx.Done())
 	}()
 
-	return resultChan, nil
+	return watchCh, nil
 }
 
-func handleWatchEvent(cfgName string, event watch.Event) clientWatchResult {
-	if event.Type == watch.Error {
-		return clientWatchResult{
-			dataStr:           "",
-			persistentContext: "",
-			err:               fmt.Errorf("error result in watcher for config '%s'", cfgName),
-		}
-	}
-
-	switch r := event.Object.(type) {
-	case *v1.Secret:
-		dataBytes, ok := r.Data[dataKeyName]
-		if !ok {
-			return clientWatchResult{
-				dataStr:           "",
-				persistentContext: "",
-				err:               fmt.Errorf("could not find data for key %s in secret %s", dataKeyName, cfgName),
+func updateHandler(kind metav1.Object, watchCh chan clientWatchResult, name string) func(prevObj interface{}, newObj interface{}) {
+	return func(prevObj, newObj interface{}) {
+		prevCast, newCast, err := cast(kind, prevObj, newObj)
+		if err != nil {
+			watchCh <- clientWatchResult{
+				err: fmt.Errorf("failed to cast objects to desired kind: %w", err),
 			}
+			return
 		}
 
-		return clientWatchResult{
-			dataStr:           string(dataBytes),
-			persistentContext: r.GetResourceVersion(),
-			err:               nil,
+		if prevCast.GetName() != newCast.GetName() {
+			// this should never happen
+			return
 		}
-	case *v1.ConfigMap:
-		dataString, ok := r.Data[dataKeyName]
-		if !ok {
-			return clientWatchResult{
-				dataStr:           "",
-				persistentContext: "",
-				err:               fmt.Errorf("could not find data for key %s in configmap %s", dataKeyName, cfgName),
-			}
+
+		if prevCast.GetName() != name {
+			// this is not the resource we want to watch, skip
+			return
 		}
 
-		return clientWatchResult{
-			dataStr:           dataString,
-			persistentContext: r.GetResourceVersion(),
-			err:               nil,
+		var newDataString string
+		var prevDataString string
+		switch kind.(type) {
+		case *v1.Secret:
+			prevDataString, newDataString, err = getSecretDataStrings(newCast, prevCast, name)
+		case *v1.ConfigMap:
+			prevDataString, newDataString, err = getConfigMapDataStrings(newCast, prevCast, name)
 		}
-	default:
-		return clientWatchResult{
-			dataStr:           "",
-			persistentContext: "",
-			err:               fmt.Errorf("unsupported type in watch %T", r),
+		watchCh <- clientWatchResult{
+			prevConfig: configRaw{
+				data:           prevDataString,
+				persistenceCtx: prevCast.GetResourceVersion(),
+			},
+			newConfig: configRaw{
+				data:           newDataString,
+				persistenceCtx: newCast.GetResourceVersion(),
+			},
+			err: err,
 		}
 	}
 }
+
+func getSecretDataStrings(newCast metav1.Object, prevCast metav1.Object, name string) (prev string, new string, err error) {
```

Comment:
Since newCast and prevCast need to be casted again in this method, I think we should not call them newCast and prevCast.